oresat_configs/scripts/gen_kaitai.py

```diff
@@ -0,0 +1,226 @@
+    canopen.objectdictionary.INTEGER8: "int8",
+    canopen.objectdictionary.INTEGER16: "i1",
+    canopen.objectdictionary.INTEGER32: "i2",
+    canopen.objectdictionary.INTEGER64: "i3",
+    canopen.objectdictionary.UNSIGNED8: "u1",
+    canopen.objectdictionary.UNSIGNED16: "u2",
+    canopen.objectdictionary.UNSIGNED32: "u3",
+    canopen.objectdictionary.UNSIGNED64: "u4",
+    canopen.objectdictionary.VISIBLE_STRING: "str",
+    canopen.objectdictionary.REAL32: "float",
+    canopen.objectdictionary.REAL64: "double",
+}
+
+
+def write_kaitai(config: OreSatConfig, dir_path: str = ".") -> None:
+    """Write beacon configs to a kaitai file."""
+
+    # Grab and format mission name
+    name = config.mission.name.lower().replace("_", ".")
+
+    #  Setup pre-determined canned types
+    kaitai_data = {
+        "meta": {
+            "id": name,
+            "title": f"{name} Decoder Struct",
+            "endian": "le",
+        },
+        "doc": "",
+        "seq": [
+            {
+                "id": "ax25_frame",
+                "type": "ax25_frame",
+                "doc-ref": "https://www.tapr.org/pub_ax25.html",
+            }
+        ],
+        "types": {
+            "ax25_frame": {
+                "seq": [
+                    {
+                        "id": "ax25_header",
+                        "type": "ax25_header",
+                    },
+                    {
+                        "id": "payload",
+                        "type": {
+                            "switch-on": "ax25_header.ctl & 0x13",
+                            "cases": {
+                                "0x03": "ui_frame",
+                                "0x13": "ui_frame",
+                                "0x00": "i_frame",
+                                "0x02": "i_frame",
+                                "0x10": "i_frame",
+                                "0x12": "i_framec",
+                            },
+                        },
+                    },
+                ]
+            },
+            "ax25_header": {
+                "seq": [
+                    {"id": "dest_callsign_raw", "type": "callsign_raw"},
+                    {"id": "dest_ssid_raw", "type": "ssid_mask"},
+                    {"id": "src_callsign_raw", "type": "callsign_raw"},
+                    {"id": "src_ssid_raw", "type": "ssid_mask"},
+                    {
+                        "id": "repeater",
+                        "type": "repeater",
+                        "if": "(src_ssid_raw.ssid_mask & 0x01) == 0",
+                        "doc": "Repeater flag is set!",
+                    },
+                    {"id": "ctl", "type": "u1"},
+                ],
+            },
+            "repeater": {
+                "seq": [
+                    {
+                        "id": "rpt_instance",
+                        "type": "repeaters",
+                        "repeat": "until",
+                        "repeat-until": "until",
```

Comment:
```
oresat0_5.ksy: /types/repeater/seq/0/repeat-until:
	error: unable to access 'until' in oresat0_5::repeater context
```
If I'm understanding things right I think this line should be `"repeat-until": "_.rpt_ssid_raw.ssid_mask & 0x01 == 1",` but I've not tested it.

Reply:
updated